File: Code Quest/Esoteric Languages/Brainfuck/runBF.js
function brainfuck(code, input = "") {
  const tape = new Array(30000).fill(0); // Memory tape with 30,000 cells
  const output = [];                     // Stores output characters
  let ptr = 0;                           // Memory pointer
  let ip = 0;                            // Instruction pointer
  let inp = 0;                           // Input character pointer
  const loopStack = [];                 // Stack to handle loop positions

  while (ip < code.length) {
    const cmd = code[ip];

    switch (cmd) {
      case '>': ptr++; break;                          // Move pointer right
      case '<': ptr--; break;                          // Move pointer left
      case '+': tape[ptr] = (tape[ptr] + 1) % 256; break; // Increment cell value
      case '-': tape[ptr] = (tape[ptr] + 255) % 256; break; // Decrement with wrap
      case '.': output.push(String.fromCharCode(tape[ptr])); break; // Output char
      case ',': tape[ptr] = input.charCodeAt(inp++) || 0; break;    // Input char

      case '[': // Start of loop
        if (tape[ptr] === 0) {
          // Skip to matching ] if cell is zero
          let open = 1;
          while (open) {
            ip++;
            if (code[ip] === '[') open++;
            if (code[ip] === ']') open--;
          }
        } else {
          loopStack.push(ip); // Store loop start
        }
        break;

      case ']': // End of loop
        if (tape[ptr] !== 0) {
          ip = loopStack[loopStack.length - 1]; // Jump back to matching [
        } else {
          loopStack.pop(); // Exit loop
        }
        break;
    }

    ip++; // Move to next instruction
  }
  return output.join(''); // Return final output string
}

// Example Usage — Uncomment to Run
// const helloWorld = "++++++++++[>+++++++>++++++++++>+++>+<<<<-]>++.>+.+++++++..+++.>++.<<+++++++++++++++.>.+++.------.--------.>+.>.";
// console.log(brainfuck(helloWorld)); // Output: Hello World!

File: Code Quest/Esoteric Languages/Brainfuck/runBF.py
def brainfuck(code, input_str=""):
    tape = [0] * 30000  # Memory tape with 30,000 cells
    output = ""         # Output string
    ptr = 0             # Memory pointer
    ip = 0              # Instruction pointer
    inp = 0             # Input pointer
    loop_stack = []     # Stack for loop tracking

    while ip < len(code):
        cmd = code[ip]

        if cmd == '>':
            ptr += 1  # Move pointer to the right
        elif cmd == '<':
            ptr -= 1  # Move pointer to the left
        elif cmd == '+':
            tape[ptr] = (tape[ptr] + 1) % 256  # Increment value at cell
        elif cmd == '-':
            tape[ptr] = (tape[ptr] - 1) % 256  # Decrement value at cell
        elif cmd == '.':
            output += chr(tape[ptr])  # Output ASCII character
        elif cmd == ',':
            # Read one character from input
            tape[ptr] = ord(input_str[inp]) if inp < len(input_str) else 0
            inp += 1

        elif cmd == '[':
            if tape[ptr] == 0:
                # Jump to the instruction after the matching ']'
                open_brackets = 1
                while open_brackets:
                    ip += 1
                    if code[ip] == '[':
                        open_brackets += 1
                    elif code[ip] == ']':
                        open_brackets -= 1
            else:
                loop_stack.append(ip)  # Save loop start index

        elif cmd == ']':
            if tape[ptr] != 0:
                ip = loop_stack[-1]  # Jump back to matching '['
            else:
                loop_stack.pop()     # Exit loop

        ip += 1  # Move to next command

    return output  # Return output string

# Example Usage — Uncomment to Run
# if __name__ == "__main__":
#    hello_world = "++++++++++[>+++++++>++++++++++>+++>+<<<<-]>++.>+.+++++++..+++.>++.<<+++++++++++++++.>.+++.------.--------.>+.>."
#    print(brainfuck(hello_world))
